chain/client/src/stateless_validation/chunk_validator/mod.rs

```diff
@@ -103,6 +107,8 @@ impl ChunkValidator {
             )));
         }
 
+        self.check_duplicate_witness(&state_witness)?;
```

Comment:
Hmm IMO this should be below `pre_validate_chunk_state_witness()`.

Otherwise a malicious chunk producer can do something like:
1. Submit valid `ChunkStateWitness`
2. Submit 100 junk `ChunkStateWitnesses` to fill the cache with junk (possibly from other random nodes to avoid getting banned)
3. Go to (1), the cache doesn't contain the valid `ChunkStateWitness` so it'll be processed again.

This way a malicious chunk producer could force the validator to process the valid `ChunkStateWitness` an arbitrary number of times.

Reply:
But could a malicious chunk producer submit 100 fake witnesses (with existing `prev_block_hash`) for the next 100 heights for which it's a chunk producer? I think we don't limit the distance from the tip for non-orphan witnesses?

Reply:
Only if we don't have a final final block for those 100 heights. If that happens then the blockchain is in the very degraded state and we have bigger problems to worry about 🙃 Keeping Flat Storage deltas in memory was built on the same assumptions by the way.

Reply:
The scenario that I'm imagining would look something like this:
There's a final block `F` at height 100
There's a chunk producer `P`, responsible for producing chunks at heights 102, 104, 106, 108, 110, ...

`P` is malicious and constructs a fake witness for each of those heights:
* Witness {height: 102, prev_block_hash: F, receipts: lotsofwork}
* Witness {height: 104, prev_block_hash: F, receipts: lotsofwork}
* Witness {height: 106, prev_block_hash: F, receipts: lotsofwork}
* ...

Then `P` sends all of those witnesses to validator `V`.

Wouldn't `V` process all of those witnesses? P could fill the whole cache with them, or just make `V` do a lot of useless work.

Reply:
Ah, I see, that is indeed a valid scenario, good catch!
Moving this PR to draft until I have a solution that addresses that.